### DISCUSSION TITLE
Hello

### DISCUSSION THREAD

**Discussion by Tobiasp1983** [syssi/esphome-smg-ii#34] [General]
Hallo, könntest du mir helfen? Ich würde gerne die Kommunikation herstellen, aber ich bekomme keine Verbindung.
Habe den IGRID2 und ESPHome installiert unter Iobroker. Flashen kann ich auch die YAML Dateien nur es kommen keine DAten.

**Comment by syssi** [comment]
Ich kenne deinen Inverter nicht. Du bist dir sicher, dass er Baugleich zum SMG ist? Wie genau hast du ESP und Inverter verdrahtet? Kannst du Fotos zur Verfügung stellen, aus welchen sich alle Details zur Verkabelung entnehmen lassen? Hast du die `uart` Komponente im YAML bereits auf gesprächig gestellt, so dass du den rohen ausgehenden und eingehenden Verkehr in Log sehen kannst?

**Reply by Tobiasp1983** [reply]
Schicke ich morgen, Ja es ist der EASUN SMG II 3,6KP 24 V. Aufgebaut habe ich alles nach deinen Bildern. 
Schicke morgen mal Bilder vom Aufbau (Ist nicht final). Ich denke da sProblem ist in ESPHome. Das ist komplett Neuland für mich. Ich hätte es am liebsten gerne das die DAten nur per MQQT gesendet werden und am besten zum IOBroker an den MQQT Client. Danke schon mal

**Reply by Tobiasp1983** [reply]
uart Komponente im YAML bereits auf gesprächig gestellt, so dass du den rohen ausgehenden und eingehenden Verkehr in Log sehen kannst? Was meinst du damit?

**Reply by syssi** [reply]
Hast du einen Linux-Rechner zur Hand, auf welchem du per `pip3 Install esphome` das PyPi-Package installierst. Auf diesem kannst du per `esphome run example.yaml` eine Konfiguration bauen, compilieren und flashen. Per `esphome log example.yaml` kannst du durchs Netzwerk aufs Logbuch der Node schauen. Hier solltest du Logmeldungen der Komponente `uart` sehen. Bei Zeilen mit dem Prefix `>>>` handelt es sich um ausgehenden Verkehr. Bei Zeilen mit dem Prefix `<<<` um eingehende Daten (in Hex).

**Comment by Tobiasp1983** [comment]
![20230724_202214](https://github.com/syssi/esphome-smg-ii/assets/140427593/04f4bdaa-2100-4947-a1ce-a15cca505c24)
![20230724_203630](https://github.com/syssi/esphome-smg-ii/assets/140427593/4665b67e-aef7-47df-9c3d-488f9803e450)

**Reply by syssi** [reply]
Dein RS232 Konverter scheint anders aufgebaut zu sein als meiner. `VCC` (am Pin-Header) bedeutet Versorgungsspannung und sollte mit 3V3 vom ESP verbunden sein. `GND` ist Masse und gehoert an `GND` vom ESP. Wie viele LEDs besitzt dein RS232-Konverter? Eine Power-LED oder eine RX- und eine TX-LED (es gibt unterschiedliche Ausfuehrungen).

**Comment by Tobiasp1983** [comment]
Ja hast du recht stecker war gerade falsch herum eingesteckt.2 Leds T1 und R3

**Reply by syssi** [reply]
Super, dann hat ein Konverter voraussichtlich eine TX- und eine RX-LED. Die TX-LED sollte periodisch blinken, weil der ESP Nachrichten in einem konstanten Intervall an den Inverter sendet. Sobald der Inverter antwortet, sollte kurz nach dem TX-Blinken auch die RX-LED aufleuchten.

**Comment by Tobiasp1983** [comment]
Also in ESP Home gehe ich auf neues Gerät-> Gebe Namen ein-> wähle ESP8266aus -> Encryption Key gehe ich auf Skip -> kopiere den Inhalt aus deiner esp8266example Yaml und flashe dies. Richtig?
Kann ich ohne geheimniss.yaml in dieser Form flashen?
![grafik](https://github.com/syssi/esphome-smg-ii/assets/140427593/64a6b02b-a314-4109-87a7-84c2ca61adb5)
Woher weis der Code welchen Port der MQQT Broker hat?

**Reply by syssi** [reply]
Genau, die `secrets.yaml` plus die `!secret`-Anweisungen in der YAML sind ein pures Suchen&Ersetzen-Makro um zu vermeiden, dass man Geheimnisse teilt, wenn man seine YAML-Konfiguration weitergibt. Du kannst also `!secret <key>`-Anweisungen entfernen und gegen das echte Geheimnis austauschen.

Viele ioBroker-Nutzer nutzen keine Authentifizierung zur Kommunikation mit ihrem Broker. In diesem Fall solltest du die Zeilen `username: xxx`  und `password: xxx` vollstaendig entfernen. Stattdessen ergaenzt du eine Zeile `port: yyyy`. Jede offizielle ESPHome-Komponente ist auf esphome.io dokumentiert. Hier findest du heraus, welche Konfigurationsparameter eine Komponente unterstuetzt/erwartet und sonst mit Defaults belegt: https://esphome.io/components/mqtt.html

**Comment by Tobiasp1983** [comment]
ESPHome client 192.168.2.140 Error: connect ECONNREFUSED 192.168.2.140:6053 das mledet der iobroker

**Comment by Tobiasp1983** [comment]
![grafik](https://github.com/syssi/esphome-smg-ii/assets/140427593/61917e97-b17f-467a-b8ca-ae11f3b4e7da)

**Comment by Tobiasp1983** [comment]
Hab es geschafft zu verbinden, und auch einmal das zumindest der MQQT Broker grün ist. Werte kommen leider noch nicht

**Reply by syssi** [reply]
Ich war den Tag lang unterwegs. Kannst du einen Logbuchauszug zur Verfügung stellen und einmal beschreiben, was die LEDs auf dem Konverter tun?

**Comment by Tobiasp1983** [comment]
Wo finde ich das Logbuch? oder benötige ich hier homeassistant?

**Reply by syssi** [reply]
Im Screenshot oben des Esphome Dashboards gibt es den Button "Log". Hier sollte ein Overlay aufgehen, in welchem man dir Logmeldungen der Node sehen kann.

**Comment by Tobiasp1983** [comment]
Beim einstecken blickt T1 für eine Zeit und der ESP ab und an mal blau blinkende. Beim Log ESP Flasher sieht man Protokolle wo er abfrägt.

**Comment by Tobiasp1983** [comment]
![grafik](https://github.com/syssi/esphome-smg-ii/assets/140427593/0ac8d100-d937-473a-a6fc-e21b8cd9b5f4)
DAs kommt. aber ich warte jetzt mal eventuell ist ja ein Bauteil defekt. Ich würde wenn das Material da ist mich nochmal melden

**Reply by syssi** [reply]
Beim Logbuch stimmt etwas nicht. Ich habe den Verdacht, dass der ESP crasht, sobald er mit dem ioBroker verbunden ist. Das passiert gerne, wenn in den MQTT-Einstellungen im ioBroker angehakt ist, dass er alle persistenten Nachrichten beim Verbindungsaufbau an den Client schicken soll. Der ESP ist mirndieser Masse an Nachrichten überfordert und streckt alle Viere von sich.

Ich würde gerne systematisch vorgehen um die Probleme einfacher identifizieren zu können. Dazu stelle ich mir das folgende Vorgehen vor:

1. Installation von ESPHome auf einem Linux-Host wie oben schon einmal erwähnt.
2. Flashen des ESPs mit entferntem `mqtt`-Abschnitt + hinzufügen einer `api:`-Zeile.
3. Punkt 2 (`api:`) hat zur Folge, dass du per `esphome log example.yaml` das Logbuch von ESPHome im Terminal angezeigt bekommst. Hier wirst du sehen, dass jede Komponente etwas zu erzählen hat. Insbesondere die `uart` Verkehr wirst du sehen.
4. Im besten Fall blinkt die TX-LED nun periodisch und pro Blink siehst du auch eine ausgehende Nachricht im Log.
5. Antwortet der Inverter? Siehst du Zeilen die mit einem `<<<` geprefixt sind? Siehst du möglicherweise sogar schon dekodierte Sensorwerte im Log? Jede Sensor-Entität wird im Log-Level `DEBUG` ausführlich über jedes Update berichtet, welche sie erfährt.
6. Antwortet der Inverter nicht, sollten wir uns noch einmal mit dem Kabel zwischen Inverter und RS232 Modul beschäftigen. Außerdem solltest du kontrollieren, ob du wirklich einen MAX3232 IC auf dem Modul hast. Es ist wichtig, dass der IC mit 3.3V Versorgungsspannung klar kommt. Es geht auch eine 5V Ausführung, welche gerne Probleme bereitet.

Sobald wir sauber mit dem Inverter kommunizieren können, hängen wir die Lösung wieder auf den MQTT-Bus und kümmern uns um die ioBroker-Konfiguration.

**Comment by Tobiasp1983** [comment]
1: Ok habe ESPHOME und Homeassistant istalliert
2. Habe den ESP geflasht nur mit dem Inhalt aus esp8266-example.yaml + MQQT Entfernt und API Key eingefügt
3. Das musst du mir genauer erläutern 
(api:) hat zur Folge, dass du per esphome log example.yaml das Logbuch von ESPHome im Terminal angezeigt bekommst. Hier wirst du sehen, dass jede Komponente etwas zu erzählen hat. Insbesondere die uart Verkehr wirst du sehen.

**Reply by syssi** [reply]
1. Home Assistant wäre nicht notwendig gewesen. Mir geht es im speziellen um das Command-Line-Tool (CLI) namens `esphome`, welches man per `pip3 install esphome` installieren kann, sofern der Python-Package-Manager installiert ist. Vielleicht hilft dir dieser Screencast um besser zu verstehen, wo ich mit dir hin will: https://raw.githubusercontent.com/syssi/esphome-seplos-bms/main/install.gif

**Comment by Tobiasp1983** [comment]
Hy jetzt zeigt er was an in Homeassistant ! Die Werte "kommen" und gehen in Status "nicht Verfügbar" wechselnd.  Nach einer WEile bleibt es bei nicht Verfügbar.
DAs Kam nach dem Eisntecken:
INFO ESPHome 2023.7.0
INFO Reading configuration /config/esphome/easun.yaml...
INFO Detected timezone 'Europe/Berlin'
INFO Starting log output from smg-ii.local using esphome API
WARNING Can't connect to ESPHome API for smg-ii.local: Error connecting to ('192.168.2.165', 6053): [Errno 111] Connect call failed ('192.168.2.165', 6053) (SocketAPIError)
INFO Trying to connect to smg-ii.local in the background
INFO Successfully connected to smg-ii.local
[11:37:51][I][app:102]: ESPHome version 2023.7.0 compiled on Jul 26 2023, 11:19:23
[11:37:51][I][app:104]: Project syssi.esphome-smg-ii version 1.0.0
[11:37:51][C][wifi:393]:   DNS1: 192.168.2.100
[11:37:51][C][logger:301]: Logger:
[11:37:51][D][uart_debug:114]: <<< 01:03:02:00:D2:38:19
[11:37:51][C][uart.arduino_esp8266:102]: UART Bus:
[11:37:51][C][uart.arduino_esp8266:103]:   TX Pin: GPIO4
[11:37:51][C][uart.arduino_esp8266:104]:   RX Pin: GPIO5
[11:37:51][C][uart.arduino_esp8266:109]:   Data Bits: 8
[11:37:51][C][uart.arduino_esp8266:111]:   Stop bits: 1
[11:37:51][D][uart_debug:114]: >>> 01:03:01:4B:00:07:75:E2
[11:37:51][D][modbus_controller.select:014]: New select value 3 from payload
[11:37:51][D][modbus_controller.sensor:025]: Sensor new state: 300.00
[11:37:51][D][sensor:093]: 'smg-ii maximum mains charging current': Sending state 10.00000 A with 1 decimals of accuracy
[11:37:51][C][modbus:143]: Modbus:
[11:37:51][D][uart_debug:114]: <<< 01:03:0E:00:03:01:2C:00:64:01:18:00:3C:00:78:00:1E:10:9E
[11:37:51][C][modbus.number:083]: modbus.numberModbus Number 'smg-ii battery equalization time'
[11:37:51][D][modbus_controller.select:014]: New select value 0 from payload
[11:37:51][C][modbus.number:083]: modbus.numberModbus Number 'smg-ii equalization Timeout exit'
[11:37:51][C][modbus.number:083]: modbus.numberModbus Number 'smg-ii two equalization charging intervals'
WARNING smg-ii.local: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for smg-ii.local
WARNING Disconnected from API
WARNING Can't connect to ESPHome API for smg-ii.local: Hello timed out (TimeoutAPIError)
INFO Trying to connect to smg-ii.local in the background
WARNING smg-ii.local: Connection error occurred: Connection lost
WARNING smg-ii.local: Connection error occurred: Connection lost
INFO Successfully connected to smg-ii.local
[11:38:40][C][modbus_controller.sensor:010]: modbus_controller.sensorModbus Controller Sensor 'smg-ii battery average Current'

**Reply by Tobiasp1983** [reply]
und nach paar minuten das:
INFO ESPHome 2023.7.0
INFO Reading configuration /config/esphome/easun.yaml...
INFO Detected timezone 'Europe/Berlin'
INFO Starting log output from smg-ii.local using esphome API
WARNING Can't connect to ESPHome API for smg-ii.local: Error connecting to ('192.168.2.165', 6053): [Errno 111] Connect call failed ('192.168.2.165', 6053) (SocketAPIError)
INFO Trying to connect to smg-ii.local in the background

**Comment by Tobiasp1983** [comment]
Ich glaube er verliert immer wieder die Verbindung , weil er immer weider Warnings meldet

WARNING Disconnected from API
WARNING Can't connect to ESPHome API for smg-ii.local: Hello timed out (TimeoutAPIError)
INFO Trying to connect to smg-ii.local in the background
WARNING smg-ii.local: Connection error occurred: Connection lost
INFO Successfully connected to smg-ii.local
[11:48:40][C][modbus_controller.sensor:010]: modbus_controller.sensorModbus Controller Sensor 'smg-ii pv average voltage'
[11:48:40][C][modbus_controller.sensor:010]: modbus_controller.sensor  Device Class: 'voltage'
[11:48:40][D][sensor:093]: 'smg-ii battery overvoltage protection point': Sending state 33.00000 V with 1 decimals of accuracy
[11:48:40][D][sensor:093]: 'smg-ii max charging voltage': Sending state 28.20000 V with 1 decimals of accuracy

**Reply by syssi** [reply]
Grundsätzlich funktioniert die Kommunikation zwischen Inverter und ESP. Hier die Rohdaten einer Anfrage + Antwort:

```
[11:37:51][D][uart_debug:114]: >>> 01:03:01:4B:00:07:75:E2
[11:37:51][D][uart_debug:114]: <<< 01:03:0E:00:03:01:2C:00:64:01:18:00:3C:00:78:00:1E:10:9E
```

Und hier kann man erkennen, dass Messwerte sauber aus den Antworten extrahieren kann:

```
[11:48:40][D][sensor:093]: 'smg-ii battery overvoltage protection point': Sending state 33.00000 V with 1 decimals of accuracy
[11:48:40][D][sensor:093]: 'smg-ii max charging voltage': Sending state 28.20000 V with 1 decimals of accuracy
```

Schlecht sind diese Zeilen:

```
WARNING smg-ii.local: Connection error occurred: Connection lost
```

Die Verbindung zum ESP ist super wackelig. Gibt es dafuer eine Erklaerung? Ist der ESP zu weit vom Accesspoint entfernt?

**Reply by Tobiasp1983** [reply]
Eigentlich nicht
Jetzt fehlt nur der MQQT von Iobroker

**Reply by syssi** [reply]
Das Thema "unstabile WLAN-Verbindung" solltest du definitiv noch angehen. Wir können aber gerne mit der MQTT-Anbindung weiter machen. Als Erstes könntest du das Log-Level senken, um etwas Ressourcen zu sparen. Ggf. läuft der ESP dann bereits ein Stück stabiler:

```
logger:
  level: INFO
```

Den `debug`-Abschnitt der `uart`-Komponente bitte auch wieder entfernen:

```
    debug:
      direction: BOTH
      dummy_receiver: false
```

Jetzt MQTT aktivieren inkl. Port:

```
mqtt:
  broker: aaa
  username: bbb
  password: ccc
  port: ddd
  id: mqtt_client
```

Und bitte einmal die Einstellungen des MQTT-Adapters in ioBroker prüfen. Sie sollten so angepasst werden:

https://github.com/syssi/esphome-jk-bms/issues/295

**Reply by Tobiasp1983** [reply]
ok werd ich zeitnah testen, ich muss dann erst iobroker zusätzlich installieren, da ich nur einen Raspi habe und a ist ja gerade der die Test Homeassist eingelegt.
Danke erstmal ich melde mich dann bei dir

**Reply by syssi** [reply]
Hast du dich um diesen Punkt gekuemmert? https://github.com/syssi/esphome-smg-ii/discussions/34#discussioncomment-6551997

Insb. die Prüfung deiner ioBroker-Konfiguration ist wichtig.

**Comment by Tobiasp1983** [comment]
er verbindet sich zware aber es kommen keine DAten an. Wenn ich den Adapter MQQT neustarte trennt er die VErbindung

INFO ESPHome 2023.7.0
INFO Reading configuration /opt/iobroker/iobroker-data/esphome.0/easun.yaml...
INFO Detected timezone 'Europe/Berlin'
INFO Starting log output from smg-ii/debug
INFO Connected to MQTT broker!
[16:28:14]
WARNING Disconnected from MQTT (7). Trying to reconnect in 1 s
WARNING Disconnected from MQTT (7). Trying to reconnect in 2 s
WARNING Disconnected from MQTT (7). Trying to reconnect in 4 s
INFO Successfully reconnected to the MQTT server
INFO Connected to MQTT broker!
[16:28:32]
WARNING Disconnected from MQTT (7). Trying to reconnect in 1 s
WARNING Disconnected from MQTT (7). Trying to reconnect in 2 s
WARNING Disconnected from MQTT (7). Trying to reconnect in 4 s
INFO Successfully reconnected to the MQTT server
INFO Connected to MQTT broker!
[16:28:46]

**Reply by syssi** [reply]
Dadurch, dass du das Log-Level auf `INFO`  gesetzt hast, siehst du auch nur noch `INFO`-Meldungen und schwerere Fehler. Wenn du wieder möchtest, dass das Log gesprächiger wird (also auch `[D]`-Meldungen gezeigt werden), musst du das Log-Level wieder anheben und den ESP neu bespielen:

``` 
logger:
  level: DEBUG
```

**Comment by Tobiasp1983** [comment]
mal ne kurze frage, geht es überhaupt an iobroker ohne esphome zu senden?

**Reply by syssi** [reply]
Ich verstehe die Frage nicht. :-) Die Komponente `mqtt` sorgt dafuer, dass Sensorwerte bei einem MQTT-Broker abgegeben werden. Die Komponente `api ` macht ein Türchen auf und wartet, dass irgendwelche interessierten Clients sich mit der ESPHome-Node verbinden. Zum Beispiel "Home Assistant" oder das esphome-Kommandozeilentool zum Abruf des Logbuchs.

**Reply by Tobiasp1983** [reply]
Ach so wenn alles läuft benötige ich API und ESPHOME etc. überhautp oder geht dein script dann nur mit Ioboker und dem MQTT.
Also wenn es mal Daten empfängt.
Also muss ich ESP überhaupt auf meinem Raspi dann haben? Ich verstehe nicht warum er verbunden ist aber keine DAten sendet bzw. Empfängt und der MqqT ach nicht "grün" wird

**Reply by syssi** [reply]
Verwerf einmal kurz alle Vorstellungen/Ideen, die du von ESPHome hast. ESPHome ist in erster Linie ein Lego-Baukasten mit ganz vielen Steinen. Die Anordnung der Steine (was wird geladen/genutzt und wie wird es verdrahtet) wid über eine YAML-Konfiguration definiert. Bittet man nun ESPHome diese YAML-Konfiguration zu interpretieren, dann wandelt ESPHome diese Beschreibung zu echtem C++-Code und danach zu einer Binärdatei, welche auf den ESP geflasht wird. Danach tut dein ESP genau das, was in der YAML beschrieben wurde. Du kannst daraufhin den Rechner runterfahren, auf welchem zuvor das Tool lag, mit welchem du die YAML interpretiert und den ESP geflasht hast. 

Wenn du von "die Node wird nicht grün" sprichst, dann sprichst du vom sog. "ESPHome Dashboard". Dabei handelt es sich um eine Tapete, dessen Aufgabe es ist alle ESPHome-Nodes im Netzwerk zu finden und 4 Buttons bereitzustellen, ueber welche du das jeweilige Logbuch der Node öffnen kannst, ggf. die YAML-Datei nacheditieren & neu flashen kannst etc. Dieses Dashboard muss man aber nicht benutzen und vor allem nicht permanent laufen haben.

**Reply by Tobiasp1983** [reply]
Ah ok .Kapiert. So mittlerweile ist er auch verbunden mit MQQT (Adaper grün) und es kommen kurz sachen an nur keine WErte, aber es geht langsam vorwärts

**Reply by Tobiasp1983** [reply]
Hy, also mittlerweile bin ich soweit fertig. Der Iobroker kommt mit den FEhlertexten nicht zurecht. WEnn ich diese Lösche, dann funtkioniert alles. Ich danke dir für die tolle Unterstützung

**Comment by Maciejdr** [comment]
Hello , Please tell me where is the problem. My inverter its SMG2, ESp32 devkitv1, and 2 kinds of RS232 converter. 
Below my Yaml file:
[smg-ii.yaml.txt](https://github.com/syssi/esphome-smg-ii/files/12704365/smg-ii.yaml.txt)

Inverter:
https://photos.google.com/photo/AF1QipPuM_cxOya4Ag90wB7IrUWe2F6AzzeilV3LF5iU
small converter:
https://photos.google.com/photo/AF1QipPye_0Zvh2vHGYg4azyMN2Hm0jv61tc2LSIcWVB
bigger converter:
https://photos.google.com/photo/AF1QipMcRnhSWFNXhruT24f3m2uK1fSrQEwOh62gPsDx
log from bigger converter:

[logs_smg-ii_logs.txt](https://github.com/syssi/esphome-smg-ii/files/12704451/logs_smg-ii_logs.txt)

log from small converterrs32:
[logs_smg-ii_logs (1).txt](https://github.com/syssi/esphome-smg-ii/files/12704453/logs_smg-ii_logs.1.txt)

Everytjing is ok, my mistake.
Led on ESp 32 and converter RS232 not blinking, but everythink works 

![Screenshot_20230924_191611_Home Assistant](https://github.com/syssi/esphome-smg-ii/assets/143712215/95f81fb6-43db-4aa4-8d83-8daa8f0dbf9d)
![Screenshot_20230924_191607_Home Assistant](https://github.com/syssi/esphome-smg-ii/assets/143712215/dedb2fc6-17b9-422a-a03c-02fd47137007)
https://photos.google.com/photo/AF1QipNtItJ1MpGMIxbXvTd06xKmAaNaIMy-ISelNiKE

**Reply by robgt978** [reply]
Can i see photo of your adapter please? My interface with esp8266 not working with this project and i want make another interface with esp32. Thank you

**Reply by syssi** [reply]
Feel free to create a new issue with some details about your ESP32 dev board. I will provide some support to identify `GPIO16, GPIO17, GND and 3V3` used here: https://github.com/syssi/esphome-smg-ii/blob/main/esp32-example.yaml
